clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1530,12 +1572,26 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     QualType ParamType = KernelParameter->getOriginalType();
     Expr *DRE = SemaRef.BuildDeclRefExpr(KernelParameter, ParamType, VK_LValue,
                                          SourceLocation());
-    if (FD->getType()->isPointerType() &&
-        FD->getType()->getPointeeType().getAddressSpace() !=
-            ParamType->getPointeeType().getAddressSpace())
-      DRE = ImplicitCastExpr::Create(SemaRef.Context, FD->getType(),
-                                     CK_AddressSpaceConversion, DRE, nullptr,
-                                     VK_RValue);
+
+    if (FD->getType()->isPointerType()) {
+      QualType ModifiedType = ParamType;
+      // Struct Type kernel arguments are decomposed. The pointer fields are
+      // then wrapped inside a compiler generated struct. Therefore when
+      // generating the initializers, we have to 'unwrap' the pointer.
+      if (MemberExprBases.size() > 2) {
```

Comment:
What will happen if we don't do this check here?

Reply:
This check restricts the wrapping to only pointers inside struct kernel arguments (i.e. only pointers which were part of a struct decomposition). It does the same thing StructDepth does in other classes.

Reply:
I think `StructDepth` looks a bit better than comparing size of the stack with "2".
We can use it here as well. But I remember that we could do this comparison several times inside the BodyCreator, so we can leave this refactoring for future commits, it is up to you.

Reply:
Thank you. I agree StructDepth looks better. I just wasn't sure whether to add that since we could technically get required information from MemberExprBases size. I will take a look at refactoring where ever this comparison occurs in BodyCreator in another commit.